unit/java_bytecode/java_bytecode_convert_method/convert_invoke_dynamic.cpp

```diff
@@ -0,0 +1,674 @@
+
+  const side_effect_exprt &side_effect_expr =
+    require_expr::require_side_effect_expr(
+      tmp_object_assignments.non_null_assignments[0].rhs(), ID_java_new);
+
+  const pointer_typet &lambda_temp_type =
+    require_type::require_pointer(side_effect_expr.type(), {});
+
+  const symbol_typet &lambda_implementor_type =
+    require_type::require_symbol(lambda_temp_type.subtype());
+
+  const irep_idt &tmp_class_identifier =
+    lambda_implementor_type.get_identifier();
+
+  const symbolt &lambda_implementor_type_symbol =
+    require_symbol::require_symbol_exists(symbol_table, tmp_class_identifier);
+
+  REQUIRE(lambda_implementor_type_symbol.is_type);
+  const class_typet &tmp_lambda_class_type =
+    require_type::require_complete_class(lambda_implementor_type_symbol.type);
+
+  REQUIRE(tmp_lambda_class_type.has_base(test_data.lambda_interface));
+  REQUIRE(tmp_lambda_class_type.has_base("java::java.lang.Object"));
+
+  class_hierarchyt class_hierarchy;
+  class_hierarchy(symbol_table);
+
+  const auto &parents = class_hierarchy.get_parents_trans(tmp_class_identifier);
+  REQUIRE_THAT(
+    parents,
+    Catch::Matchers::Vector::ContainsElementMatcher<irep_idt>{
+      test_data.lambda_interface});
+
+  const auto &interface_children =
+    class_hierarchy.get_children_trans(test_data.lambda_interface);
+
+  REQUIRE_THAT(
+    interface_children,
+    Catch::Matchers::Vector::ContainsElementMatcher<irep_idt>{
+      tmp_class_identifier});
+
+  const java_class_typet::componentt super_class_component =
+    require_type::require_component(tmp_lambda_class_type, "@java.lang.Object");
+
+  const symbol_typet &super_class_type = require_type::require_symbol(
+    super_class_component.type(), "java::java.lang.Object");
+
+  const symbolt &base_class_symbol = require_symbol::require_symbol_exists(
+    symbol_table, super_class_type.get_identifier());
+
+  REQUIRE(base_class_symbol.is_type);
+  const class_typet &super_class_type_struct =
+    require_type::require_incomplete_class(base_class_symbol.type);
+
+  require_type::require_component(super_class_type_struct, "@class_identifier");
+  // TODO verify the components of the class have been set correctly
+
+  THEN("The function in the class should call the lambda method")
+  {
+    const irep_idt method_identifier =
+      id2string(tmp_class_identifier) +
+      test_data.lambda_interface_method_descriptor;
+    const symbolt &method_symbol =
+      require_symbol::require_symbol_exists(symbol_table, method_identifier);
+
+    REQUIRE(method_symbol.is_function());
+
+    const std::vector<codet> &assignments =
+      require_goto_statements::get_all_statements(method_symbol.value);
+
+    std::vector<code_function_callt> function_calls =
+      require_goto_statements::find_function_calls(
+        assignments, test_data.lambda_function_id);
+
+    INFO("Looking for function call of " << test_data.lambda_function_id);
+    REQUIRE(function_calls.size() == 1);
```

Comment:
Lambda function can be called more than once in a method, isn't this too restrictive? Or does each lambda call get it's own object?

Reply:
This was discussed offline, there indeed should be exactly one call to the lambda from the code we generate.